packages/@aws-cdk/aws-apigateway/lib/model.ts

```diff
@@ -1,46 +1,141 @@
-export interface IModel {
-    readonly modelId: string;
-}
+import { Construct, PhysicalName, Resource } from '@aws-cdk/cdk';
+import { CfnModel, CfnModelProps } from './apigateway.generated';
+import jsonSchema = require('./json-schema');
+import { IRestApi, RestApi } from './restapi';
 
-/**
- * Represents a reference to a REST API's Empty model, which is available
- * as part of the model collection by default. This can be used for mapping
- * JSON responses from an integration to what is returned to a client,
- * where strong typing is not required. In the absence of any defined
- * model, the Empty model will be used to return the response payload
- * unmapped.
- *
- * Definition
- * {
- *   "$schema" : "http://json-schema.org/draft-04/schema#",
- *   "title" : "Empty Schema",
- *   "type" : "object"
- * }
- *
- * @see https://docs.amazonaws.cn/en_us/apigateway/latest/developerguide/models-mappings.html#models-mappings-models
- */
-export class EmptyModel implements IModel {
-    public readonly modelId = 'Empty';
+export interface IModelRef {
+  /** @attribute */
+  readonly modelName: string;
 }
 
-/**
- * Represents a reference to a REST API's Error model, which is available
- * as part of the model collection by default. This can be used for mapping
- * error JSON responses from an integration to a client, where a simple
- * generic message field is sufficient to map and return an error payload.
- *
- * Definition
- * {
- *   "$schema" : "http://json-schema.org/draft-04/schema#",
- *   "title" : "Error Schema",
- *   "type" : "object",
- *   "properties" : {
- *     "message" : { "type" : "string" }
- *   }
- * }
- */
-export class ErrorModel implements IModel {
-    public readonly modelId = 'Error';
+export interface ModelProps {
+  /**
+   * The rest API that this model is part of.
+   *
+   * The reason we need the RestApi object itself and not just the ID is because the model
+   * is being tracked by the top-level RestApi object for the purpose of calculating it's
+   * hash to determine the ID of the deployment. This allows us to automatically update
+   * the deployment when the model of the REST API changes.
+   *
+   * @default None: This is automatically populated by the api.addRequestValidator method
+   */
+  readonly restApi?: IRestApi;
```

Comment:
Why is this optional? You can extract all props by this to `ModelOptions` (like we do with `MethodOptions`) and represent the fact that this is required when defining a `Model` class directly.

Reply:
Moved to a ModelOptions and ModelProps syntax, where ModelProps is an extension of ModelOptions with the defined restApi property as required for readability (I have seen that in some classes). The convenience method (addModel) expects a ModelOptions as a parameter.